File: ScalaQuickStart/src/main/java/com/tirthal/learning/java2scala/javaway/controlstruct/LoopsBreaksSyntax.java
package com.tirthal.learning.java2scala.javaway.controlstruct;

import java.util.Arrays;
import java.util.List;

// Syntax of loops and breaks control structures in Java
public class LoopsBreaksSyntax {

	public static void main(String[] args) {		

		// (1) ---> while loop  
		int i = 0;
		while(i < 3) {
			System.out.println(i);
			i++;
		}
		
		// (2) ---> for loop = initialize, check and update
		for(int j = 0; j < i; j++) {
			System.out.println(j);
		}
		
		// for each
		List<Integer> numbers = Arrays.asList(0, 1, 2, 3);
		numbers.forEach(n -> System.out.println(n)); // Alternate syntax = numbers.forEach(System.out::println);
		
		// (3) ---> break
		for(int k = 0; k < 10; k++) {
			System.out.println(k);
			if(k == 5)
				break;			
		}
	}

}

File: ScalaQuickStart/src/main/scala/com/tirthal/learning/java2scala/scalaway/controlstruct/LoopsBreaksSyntax.scala
package com.tirthal.learning.java2scala.scalaway.controlstruct

// Syntax of loops and breaks control structures in Scala
object LoopsBreaksSyntax {

  def main(args: Array[String]) {
    
    // (1) ---> while loop
    var i: Int = 0
    while (i < 3) {
      println(i)
      i += 1 // there is no i++ operator in scala
    } 
    
    // (2) ---> Scala has for comprehension (instead of for loop)
    for(i <- 0 to i) { // "<-" indicates a generator to feed value into the iterator
      println(i)
    }
    
    // for each
    (0 to 3).foreach(i => println(i)) // Alternate syntax of using Scala's shorthand = (0 to 3).foreach(println(_))
    
    // (3) ---> Scala has break() API instead of break statement
    // Scala has no break/continue statements, because it generally discourages to breaking of loops
    import scala.util.control.Breaks._ // Scala allows to import within method block also
    breakable {
      for(k <- 0 to 10) {
        println(k)
        if(i ==5)
          break()
      }
    }
  }
}